<a href="https://colab.research.google.com/github/Olga-Lyt/DTA-2026/blob/main/HW/Python_practice_ipynb%22.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 🐍 Python: Практикум

**Час виконання:** 3 години  
**Формат:** індивідуальна робота

---

## 📋 Що ви будете практикувати

| Розділ | Тема | Кількість завдань |
|---|---|---|
| 1 | `print`, `input`, змінні | 8 |
| 2 | `if / elif / else` (умови) | 10 |
| 3 | Цикл `for` з `range()` | 10 |
| 4 | Цикл `for` з рядками і списками | 10 |
| 5 | Цикл `while` | 8 |
| 6 | Списки (`list`) — методи та операції | 10 |
| 7 | 🔥 Комбіновані завдання | 6 |

**Усього:** 62 завдання

---

## 📌 Як працювати з зошитом

1. Читайте умову уважно — там може бути більше деталей, ніж здається
2. Не поспішайте — якщо завдання здається складним, почніть з простого варіанту, а потім ускладнюйте
3. Запускайте код часто (Shift + Enter), щоб перевіряти, що все працює
4. Якщо щось не виходить — перечитайте підказку (якщо є), погугліть, поекспериментуйте
5. **Помилки — це нормально!** Кожен програміст бачить їх сотні разів на день

> 💡 У завданнях позначено: ⭐ — легке, ⭐⭐ — середнє, ⭐⭐⭐ — складне

---

# Розділ 1. `print`, `input` та змінні

Базові завдання на введення-виведення та роботу зі змінними.

**Що треба знати:**
- `print(...)` — вивести на екран
- `input("текст: ")` — отримати ввід від користувача (повертає текст!)
- `int(...)`, `float(...)` — перетворити текст у число
- Арифметика: `+`, `-`, `*`, `/`, `**` (степінь), `%` (залишок)

### Завдання 1. Привітання ⭐

Запитайте у користувача його ім'я і виведіть привітання.

**Приклад:**
```
Як вас звати? Олена
Привіт, Олена!
```

In [ ]:
# Ваш код для завдання 1:
name = input("Як вас звати? ")
print("Привіт,", name + "!")


Як вас звати? Олена
Привіт, Олена!


### Завдання 2. Сума двох чисел ⭐

Запитайте у користувача два числа і виведіть їхню суму.

**Приклад:**
```
Перше число: 7
Друге число: 12
Сума: 19
```

> 💡 **Підказка:** Не забудьте перетворити введення на число через `int(...)`.

In [ ]:
# Ваш код для завдання 2:
x = int(input("Введіть перше число "))
y = int(input("Введіть друге число "))
print("Сума чисел:", x + y)

Введіть перше число 7
Введіть друге число 12
Сума чисел: 19


### Завдання 3. Площа прямокутника ⭐

Запитайте довжину та ширину прямокутника і виведіть його площу.

In [ ]:
# Ваш код для завдання 3:
x = int(input("Довжина прямокутника "))
y = int(input("Ширина прямокутника "))
print("Площа прямокутника:", x * y)


Довжина прямокутника 8
Ширина прямокутника 7
Площа прямокутника: 56


### Завдання 4. Сантиметри в метри ⭐

Запитайте у користувача кількість сантиметрів і переведіть їх у метри.

**Приклад:**
```
Кількість см: 250
Це 2.5 м
```

In [ ]:
x = int(input("Кількість см: "))
m = x / 100
print(f"Це {m} м")

Кількість см: 250
Це 2.5 м


### Завдання 5. Округлення оцінки ⭐⭐

Запитайте у користувача оцінку (число з крапкою, наприклад 4.7).
Виведіть її, округлену до цілого числа.

> 💡 **Підказка:** Використайте функцію `round()`. Не забудьте `float()` замість `int()`.

In [ ]:
# Ваш код для завдання 5:
x = float(input("Введіть оцінку"))
print(round(x))

Введіть оцінку4.7
5


### Завдання 6. Ціна з ПДВ ⭐⭐

Запитайте у користувача ціну товару без ПДВ.
Розрахуйте та виведіть:
- суму ПДВ (20% від ціни)
- кінцеву ціну (ціна + ПДВ)

In [ ]:
x = float(input("Ціна без ПДВ: "))
vat_amount = x * 0.20  # 20% ПДВ
final_price = x + vat_amount
print(f"Сума ПДВ: {vat_amount}")
print(f"Кінцева ціна: {final_price}")

Ціна без ПДВ: 52.7
Сума ПДВ: 10.540000000000001
Кінцева ціна: 63.24


### Завдання 7. Зворотній відлік ⭐⭐

Запитайте у користувача число `n` і виведіть на одному рядку числа від `n` до 1, розділені пробілами.

**Приклад:**
```
Введіть число: 5
5 4 3 2 1
```

> 💡 **Підказка:** Можна використати `print(n, n-1, n-2, n-3, n-4)` — лише для цього завдання, без циклу.

In [ ]:
# Ваш код для завдання 7:
n = int(input("Введіть число: "))
print(n, n-1, n-2, n-3, n-4)

Введіть число: 5
5 4 3 2 1


### Завдання 8. Обмін значень змінних ⭐⭐⭐

Створіть дві змінні `a = 10` і `b = 20`. Поміняйте їхні значення місцями (щоб `a` стало 20, а `b` — 10) і виведіть результат.

> 💡 **Підказка:** Можна через третю тимчасову змінну, або в Python є дуже короткий спосіб: `a, b = b, a`.

In [ ]:
# Ваш код для завдання 8:
a = 10
b = 20
a, b = b, a
print(a, b)

20 10


---

# Розділ 2. Умови `if / elif / else`

**Що треба знати:**
- `if умова:` — якщо умова правдива
- `elif умова:` — інакше якщо
- `else:` — у всіх інших випадках
- Оператори порівняння: `==`, `!=`, `>`, `<`, `>=`, `<=`
- Логічні оператори: `and`, `or`, `not`

### Завдання 9. Парне чи непарне ⭐

Запитайте у користувача число і виведіть, парне воно чи непарне.

> 💡 **Підказка:** Парне число ділиться на 2 без остачі: `n % 2 == 0`.

In [ ]:
# Ваш код для завдання 9:
x = int(input("Введіть число "))
if x % 2 == 0:
    print("Число парне")
else:
    print("Число непарне")

Введіть число 7
Число непарне


### Завдання 10. Додатнє, від'ємне чи нуль ⭐

Запитайте у користувача число і виведіть, додатнє воно, від'ємне чи дорівнює нулю.

In [ ]:
# Ваш код для завдання 10:
x = float(input("Введіть число "))
if x < 0:
    print("Число від'ємне")
elif x == 0:
    print("Дорівнює нулю")
else:
    print("Число додатнє")

Введіть число 15
Число додатнє


### Завдання 11. Більше з двох ⭐

Запитайте у користувача два числа і виведіть більше з них.
Якщо числа рівні — виведіть «числа рівні».

In [ ]:
# Ваш код для завдання 11:
x = float(input("Введіть перше число "))
y = float(input("Введіть друге число "))
if x < y:
    print(y)
elif x == y:
    print("Числа рівні")
else:
    print(x)

Введіть перше число 27
Введіть друге число 63
63.0


### Завдання 12. Чи може людина голосувати ⭐

Запитайте у користувача вік. Виведіть:
- «Ви можете голосувати» — якщо вік ≥ 18
- «Ви ще малі для голосування» — якщо вік < 18
- «Некоректний вік» — якщо вік від'ємний

In [ ]:
# Ваш код для завдання 12:
x = int(input("Скільки Вам років? "))
if x < 0:
    print("Некоректний вік")
elif x >= 18:
    print("Ви можете голосувати")
else:
    print("Ви ще малі для голосування")

Скільки Вам років? 32
Ви можете голосувати


### Завдання 13. Оцінка за бал ⭐⭐

Запитайте у користувача оцінку (від 0 до 100) і виведіть:
- 90+ → «Відмінно»
- 75-89 → «Добре»
- 60-74 → «Задовільно»
- 0-59 → «Не склав»
- інакше (від'ємні або >100) → «Некоректна оцінка»

In [ ]:
# Ваш код для завдання 13:
score = int(input("Введіть оцінку (від 0 до 100): "))

if score < 0 or score > 100:
    print("Некоректна оцінка")
elif score >= 90:
    print("Відмінно")
elif score >= 75:
    print("Добре")
elif score >= 60:
    print("Задовільно")
else:
    print("Не склав")

Введіть оцінку (від 0 до 100): 86
Добре


### Завдання 14. Найбільше з трьох ⭐⭐

Запитайте у користувача три числа і виведіть найбільше з них.

> 💡 **Підказка:** Можна або послідовно через `if/elif`, або скористатись вбудованою функцією `max()`.

In [ ]:
# Ваш код для завдання 14:
x = float(input("Введіть перше число: "))
y = float(input("Введіть друге число: "))
z = float(input("Введіть третє число: "))

maximum_number = max(x, y, z)
print(f"Найбільше число: {maximum_number}")

Введіть перше число: 16
Введіть друге число: 12
Введіть третє число: 14
Найбільше число: 16.0


### Завдання 15. Високосний рік ⭐⭐

Запитайте у користувача рік і визначте, чи є він високосним.

**Правила:** Рік є високосним, якщо:
- ділиться на 4, АЛЕ
- НЕ ділиться на 100, АБО
- ділиться на 400

**Приклади:** 2024 — високосний, 2023 — ні, 1900 — ні, 2000 — високосний.

> 💡 **Підказка:** Використайте логічні оператори `and` та `or`. Умова: `(рік % 4 == 0 and рік % 100 != 0) or рік % 400 == 0`.

In [ ]:
# Ваш код для завдання 15:
year = int(input("Введіть рік: "))

if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
    print(f"{year} рік є високосним")
else:
    print(f"{year} рік не є високосним")

Введіть рік: 2032
2032 рік є високосним


### Завдання 16. Пори року ⭐⭐

Запитайте у користувача номер місяця (1-12) і виведіть, до якої пори року він належить:
- 12, 1, 2 → зима
- 3, 4, 5 → весна
- 6, 7, 8 → літо
- 9, 10, 11 → осінь
- інакше → «Некоректний місяць»

In [ ]:
# Ваш код для завдання 16:
month = int(input("Введіть номер місяця (1-12): "))

if month == 12 or 1 <= month <= 2:
    print("Зима")
elif 3 <= month <= 5:
    print("Весна")
elif 6 <= month <= 8:
    print("Літо")
elif 9 <= month <= 11:
    print("Осінь")
else:
    print("Некоректний місяць")

Введіть номер місяця (1-12): 10
Осінь


### Завдання 17. Калькулятор ⭐⭐

Запитайте у користувача два числа і знак операції (`+`, `-`, `*`, `/`).
Виведіть результат відповідної операції.

**Приклад:**
```
Перше число: 10
Друге число: 3
Знак: *
Результат: 30
```

⚠️ Не забудьте перевірити ділення на 0!

In [ ]:
# Ваш код для завдання 17:
num1 = float(input("Введіть перше число: "))
num2 = float(input("Введіть друге число: "))
operator = input("Введіть знак операції (+, -, *, /): ")

if operator == '+':
    result = num1 + num2
elif operator == '-':
    result = num1 - num2
elif operator == '*':
    result = num1 * num2
elif operator == '/':
    if num2 != 0:
        result = num1 / num2
    else:
        result = "Помилка: ділення на нуль!"
else:
    result = "Некоректний знак операції"

print(f"Результат: {result}")

Введіть перше число: 10
Введіть друге число: 3
Введіть знак операції (+, -, *, /): *
Результат: 30.0


### Завдання 18. Трикутник ⭐⭐⭐

Запитайте у користувача три довжини сторін і визначте:
1. Чи можна з них скласти трикутник? (правило: сума будь-яких двох сторін має бути більшою за третю)
2. Якщо можна — який це трикутник:
   - рівносторонній (всі сторони рівні)
   - рівнобедрений (дві сторони рівні)
   - різносторонній (всі різні)

> 💡 **Підказка:** Спочатку перевірте, чи трикутник взагалі існує. Потім вже визначайте тип через порівняння сторін.

In [ ]:
a = float(input("Введіть довжину першої сторони: "))
b = float(input("Введіть довжину другої сторони: "))
c = float(input("Введіть довжину третьої сторони: "))

if (a + b > c) and (a + c > b) and (b + c > a):
    print("З цих сторін можна скласти трикутник.")
    if a == b and b == c:
        print("Це рівносторонній трикутник.")
    elif a == b or b == c or a == c:
        print("Це рівнобедрений трикутник.")
    else:
        print("Це різносторонній трикутник.")
else:
    print("З цих сторін не можна скласти трикутник.")

Введіть довжину першої сторони: 12
Введіть довжину другої сторони: 12
Введіть довжину третьої сторони: 25
З цих сторін не можна скласти трикутник.


---

# Розділ 3. Цикл `for` з `range()`

**Що треба знати:**
- `range(n)` → 0, 1, ..., n-1
- `range(a, b)` → a, a+1, ..., b-1
- `range(a, b, step)` → a, a+step, ... (з кроком, не доходячи до b)

### Завдання 19. Числа від 1 до 20 ⭐

Виведіть всі числа від 1 до 20, кожне на новому рядку.

In [ ]:
# Ваш код для завдання 19:
for i in range(1, 21):
    print(i)

1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20


### Завдання 20. Парні числа ⭐

Виведіть всі парні числа від 2 до 30.

> 💡 **Підказка:** Можна через `range(2, 31, 2)` або через перевірку `if n % 2 == 0`.

In [ ]:
# Ваш код для завдання 20:
for i in range(2, 31, 2):
    print(i)

2
4
6
8
10
12
14
16
18
20
22
24
26
28
30


### Завдання 21. Сума від 1 до 100 ⭐

Підрахуйте суму всіх чисел від 1 до 100 і виведіть результат. Очікуване число — 5050.

In [ ]:
# Ваш код для завдання 21:
total_sum = 0
for i in range(1, 101):
    total_sum += i
print(f"Сума чисел від 1 до 100: {total_sum}")

Сума чисел від 1 до 100: 5050


### Завдання 22. Таблиця множення ⭐

Запитайте у користувача число і виведіть таблицю множення цього числа на числа від 1 до 10.

**Приклад при введенні 7:**
```
7 * 1 = 7
7 * 2 = 14
...
7 * 10 = 70
```

In [ ]:
# Ваш код для завдання 22:
num = int(input("Введіть число: "))

for i in range(1, 11):
    print(f"{num} * {i} = {num * i}")

Введіть число: 7
7 * 1 = 7
7 * 2 = 14
7 * 3 = 21
7 * 4 = 28
7 * 5 = 35
7 * 6 = 42
7 * 7 = 49
7 * 8 = 56
7 * 9 = 63
7 * 10 = 70


### Завдання 23. Сума парних чисел ⭐⭐

Підрахуйте суму всіх парних чисел від 1 до 100.

In [ ]:
total_even_sum = 0
for i in range(2, 101, 2):
    total_even_sum += i
print(f"Сума всіх парних чисел від 1 до 100: {total_even_sum}")

Сума всіх парних чисел від 1 до 100: 2550


### Завдання 24. Факторіал ⭐⭐

Запитайте у користувача число `n` і обчисліть його факторіал (`n!`).

**Нагадування:** факторіал — це добуток всіх чисел від 1 до `n`.  
Наприклад, `5! = 1 * 2 * 3 * 4 * 5 = 120`.

> 💡 **Підказка:** Створіть змінну `result = 1`, а потім у циклі множте її на кожне число.

In [ ]:
n = int(input("Введіть число для обчислення факторіалу: "))

if n < 0:
    print("Факторіал не визначений для від'ємних чисел.")
elif n == 0:
    print("Факторіал 0 дорівнює 1.")
else:
    factorial = 1
    for i in range(1, n + 1):
        factorial *= i
    print(f"Факторіал числа {n} дорівнює {factorial}.")

Введіть число для обчислення факторіалу: 5
Факторіал числа 5 дорівнює 120.


### Завдання 25. Степінь числа ⭐⭐

Запитайте у користувача два числа: основу `a` і степінь `n`.
Обчисліть `a` в степені `n`, НЕ використовуючи оператор `**`.

**Приклад:** `a = 3, n = 4` → `3 * 3 * 3 * 3 = 81`

> 💡 **Підказка:** Створіть змінну `result = 1` і помножте її на `a` рівно `n` разів.

In [ ]:
a = float(input("Введіть основу (число a): "))
n = int(input("Введіть степінь (ціле число n): "))

result = 1

if n < 0:
    print("Для цього завдання степінь має бути невід'ємним цілим числом.")
elif n == 0:
    result = 1
else:
    for _ in range(n):
        result *= a

print(f"Число {a} в степені {n} дорівнює {result}")

Введіть основу (число a): 7
Введіть степінь (ціле число n): 4
Число 7.0 в степені 4 дорівнює 2401.0


### Завдання 26. Сума цифр числа ⭐⭐⭐

Запитайте у користувача число і знайдіть суму його цифр.

**Приклад:** число 1234 → сума 1+2+3+4 = 10

> 💡 **Підказка:** Найпростіший спосіб: перетворити число на рядок через `str()`, а потім перебрати символи у циклі і додавати `int(символ)` до суми.

In [ ]:
num_str = input("Введіть число: ")

total_sum_digits = 0
for digit_char in num_str:
    if digit_char.isdigit():
        total_sum_digits += int(digit_char)
    else:
        print("Будь ласка, вводьте лише цифри.")
        total_sum_digits = -1
        break

if total_sum_digits != -1:
    print(f"Сума цифр числа {num_str} дорівнює {total_sum_digits}.")

Введіть число: 12345
Сума цифр числа 12345 дорівнює 15.


### Завдання 27. Прості числа ⭐⭐⭐

Виведіть всі прості числа від 2 до 50.

**Нагадування:** просте число — це число, яке ділиться тільки на 1 і саме на себе.

> 💡 **Підказка:** Для кожного числа `n` від 2 до 50 перевіряйте, чи має воно дільники від 2 до n-1. Можна використати змінну-прапорець `is_prime = True`.

In [ ]:
# Ваш код для завдання 27:
for num in range(2, 51):
    is_prime = True
    for i in range(2, num):
        if num % i == 0:
            is_prime = False
            break
    if is_prime:
        print(num)

2
3
5
7
11
13
17
19
23
29
31
37
41
43
47


### Завдання 28. Малювання трикутника ⭐⭐⭐

Виведіть на екран прямокутний трикутник з зірочок. Висоту запитайте у користувача.

**Приклад при висоті 5:**
```
*
**
***
****
*****
```

> 💡 **Підказка:** Використайте вкладений цикл АБО операцію множення рядка: `'*' * n` повертає рядок з `n` зірочок.

In [ ]:
# Ваш код для завдання 28:
height = int(input("Введіть висоту трикутника: "))

for i in range(1, height + 1):
    print('*' * i)

Введіть висоту трикутника: 7
*
**
***
****
*****
******
*******


---

# Розділ 4. Цикл `for` з рядками та списками

**Що треба знати:**
- `for letter in "Python":` — перебираємо літери в рядку
- `for item in [1, 2, 3]:` — перебираємо елементи списку
- `len(x)` — довжина рядка або списку
- Перевірка наявності: `if "a" in word:`

### Завдання 29. Довжина слова ⭐

Запитайте у користувача слово і виведіть кількість літер у ньому.

> 💡 **Підказка:** Використайте функцію `len()`.

In [ ]:
word = input("Введіть слово: ")
print(f"Довжина слова: {len(word)}")

Введіть слово: завдання
Довжина слова: 8


### Завдання 30. Виведення кожної літери ⭐

Запитайте у користувача слово і виведіть кожну літеру на новому рядку.

In [ ]:
word = input("Введіть слово: ")

for letter in word:
    print(letter)

Введіть слово: завдання
з
а
в
д
а
н
н
я


### Завдання 31. Підрахунок голосних ⭐⭐

Запитайте у користувача слово і підрахуйте кількість голосних літер у ньому.

Голосні: `а е и і о у я ю є ї`.

> 💡 **Підказка:** Створіть лічильник `count = 0`. У циклі перевіряйте: `if letter in "аеиіоуяюєї":` і збільшуйте лічильник.

In [ ]:
word = input("Введіть слово: ")
vowels = "аеиіоуяюєїАЕИІОУЯЮЄЇ"
vowel_count = 0

for letter in word:
    if letter in vowels:
        vowel_count += 1

print(f"Кількість голосних літер у слові '{word}': {vowel_count}")

Введіть слово: завдання
Кількість голосних літер у слові 'завдання': 3


### Завдання 32. Перевернуте слово ⭐⭐

Запитайте у користувача слово і виведіть його у зворотному порядку.

**Приклад:** `Python` → `nohtyP`

> 💡 **Підказка:** Створіть порожній рядок `result = ''` і у циклі додавайте кожну літеру НА ПОЧАТОК: `result = letter + result`.

In [ ]:
word = input("Введіть слово: ")
reversed_word = ""

for letter in word:
    reversed_word = letter + reversed_word

print(reversed_word)

Введіть слово: завдання
яннадваз


### Завдання 33. Підрахунок певної літери ⭐⭐

Запитайте у користувача слово і літеру. Виведіть, скільки разів ця літера зустрічається у слові.

In [ ]:
word = input("Введіть слово: ")
target_letter = input("Введіть літеру для підрахунку: ")

count = 0
for letter in word:
    if letter == target_letter:
        count += 1

print(f"Літера '{target_letter}' зустрічається у слові '{word}' {count} раз(ів).")

Введіть слово: завдання
Введіть літеру для підрахунку: а
Літера 'а' зустрічається у слові 'завдання' 2 раз(ів).


### Завдання 34. Сума елементів списку ⭐

Дано список `numbers = [12, 7, 18, 3, 25, 9, 14]`. Знайдіть і виведіть суму всіх його елементів. Очікуваний результат: 88.

In [ ]:
numbers = [12, 7, 18, 3, 25, 9, 14]
total_sum = 0
for num in numbers:
    total_sum += num
print(f"Сума елементів списку: {total_sum}")

Сума елементів списку: 88


### Завдання 35. Найбільший і найменший ⭐⭐

Дано список `numbers = [42, 17, 89, 3, 56, 91, 24, 8, 73]`. Знайдіть найбільший і найменший елементи **без використання `max()` та `min()`**.

> 💡 **Підказка:** Створіть змінні `maximum = numbers[0]` і `minimum = numbers[0]`, а потім у циклі порівнюйте.

In [ ]:
numbers = [42, 17, 89, 3, 56, 91, 24, 8, 73]

maximum = numbers[0]
minimum = numbers[0]

for num in numbers:
    if num > maximum:
        maximum = num
    if num < minimum:
        minimum = num

print(f"Найбільший елемент: {maximum}")
print(f"Найменший елемент: {minimum}")

Найбільший елемент: 91
Найменший елемент: 3


### Завдання 36. Тільки додатні ⭐⭐

Дано список з додатними та від'ємними числами. Виведіть лише додатні.

In [ ]:
numbers = [12, -5, 7, -18, 25, 0, -3, 9, -14, 30]
print("Додатні числа:")
for num in numbers:
    if num > 0:
        print(num)

Додатні числа:
12
7
25
9
30


### Завдання 37. Середній бал ⭐⭐

Дано список оцінок студентів. Підрахуйте середній бал.

> 💡 **Підказка:** Середнє = сума / кількість. Кількість можна отримати через `len()`.

In [ ]:
grades = [85, 72, 90, 65, 78, 88, 92, 70, 81, 76]
total_grades = sum(grades)
num_grades = len(grades)
average_grade = total_grades / num_grades
print(f"Середній бал: {average_grade:.2f}")

Середній бал: 79.70


### Завдання 38. Найдовше слово ⭐⭐⭐

Дано список слів. Знайдіть і виведіть найдовше слово.

> 💡 **Підказка:** Створіть змінну `longest = words[0]`, а потім у циклі порівнюйте через `len()`.

In [ ]:
words = ["кіт", "програмування", "Python", "школа", "інформатика", "олівець"]

longest_word = words[0]

for word in words:
    if len(word) > len(longest_word):
        longest_word = word

print(f"Найдовше слово у списку: {longest_word}")

Найдовше слово у списку: програмування


---

# Розділ 5. Цикл `while`

**Що треба знати:**
- `while умова:` — повторюємо, поки умова правдива
- ⚠️ Завжди має бути щось, що змінює умову — інакше нескінченний цикл!
- Використовуємо `while`, коли наперед не відомо, скільки разів повторювати

### Завдання 39. Лічильник до 10 ⭐

За допомогою циклу `while` виведіть числа від 1 до 10.

In [ ]:
# Ваш код для завдання 39:
count = 1
while count <= 10:
    print(count)
    count += 1

1
2
3
4
5
6
7
8
9
10


### Завдання 40. Зворотній відлік ⭐

За допомогою циклу `while` виведіть числа від 10 до 1, потім надрукуйте «Старт!».

In [ ]:
# Ваш код для завдання 40:
count = 10
while count >= 1:
    print(count)
    count -= 1
print("Старт!")

10
9
8
7
6
5
4
3
2
1
Старт!


### Завдання 41. Подвоєння ⭐⭐

Починаючи з числа 1, подвоюйте його, поки воно менше 1000. Виводьте кожне значення.

**Очікуваний результат:** 1, 2, 4, 8, 16, 32, 64, 128, 256, 512.

In [ ]:
# Ваш код для завдання 41:
number = 1
while number < 1000:
    print(number)
    number *= 2

1
2
4
8
16
32
64
128
256
512


### Завдання 42. Сума введених чисел ⭐⭐

Просіть користувача вводити числа, доки він не введе 0.
Після цього виведіть суму всіх введених чисел (нуль не враховується).

> 💡 **Підказка:** Створіть змінну `total = 0` і змінну `number = -1` (щоб умова `while number != 0:` спрацювала на початку). Усередині циклу запитуйте число і додавайте до суми.

In [ ]:
total = 0
number = -1
while number != 0:
    number = int(input("Введіть число (0 для завершення): "))
    if number != 0:
        total += number

print(f"Сума всіх введених чисел: {total}")

Введіть число (0 для завершення): 12
Введіть число (0 для завершення): 7
Введіть число (0 для завершення): 32
Введіть число (0 для завершення): 0
Сума всіх введених чисел: 51


### Завдання 43. Кількість цифр у числі ⭐⭐

Запитайте у користувача додатнє ціле число і підрахуйте, скільки в ньому цифр.

**Приклад:** число 12345 → 5 цифр.

> 💡 **Підказка:** Поки число більше за 0, ділимо його націло на 10 (`n = n // 10`) і збільшуємо лічильник на 1.

In [ ]:
num = int(input("Введіть додатнє ціле число: "))

if num < 0:
    print("Будь ласка, введіть додатнє ціле число.")
elif num == 0:
    print("Число 0 має 1 цифру.")
else:
    count = 0
    temp_num = num
    while temp_num > 0:
        temp_num = temp_num // 10
        count += 1
    print(f"Число {num} має {count} цифр.")

Введіть додатнє ціле число: 1234567
Число 1234567 має 7 цифр.


### Завдання 44. Перевірка пароля ⭐⭐

Створіть змінну `correct_password = "qwerty123"`.
У циклі `while` запитуйте у користувача пароль, доки він не введе правильний.
Після успішного введення виведіть «Доступ дозволено».

In [ ]:
correct_password = "qwerty123"
user_password = ""

while user_password != correct_password:
    user_password = input("Введіть пароль: ")
    if user_password == correct_password:
        print("Доступ дозволено")
    else:
        print("Неправильний пароль. Спробуйте ще раз.")

Введіть пароль: qwerty
Неправильний пароль. Спробуйте ще раз.
Введіть пароль: 123
Неправильний пароль. Спробуйте ще раз.
Введіть пароль: qwerty123
Доступ дозволено


### Завдання 45. Сума цифр через while ⭐⭐⭐

Запитайте у користувача додатнє ціле число і знайдіть суму його цифр, використовуючи `while`.

**Приклад:** число 1234 → сума 1+2+3+4 = 10.

> 💡 **Підказка:** У циклі: беремо останню цифру через `n % 10`, додаємо до суми, потім відкидаємо її через `n = n // 10`. Цикл працює, поки `n > 0`.

In [ ]:
num = int(input("Введіть додатнє ціле число: "))

if num < 0:
    print("Будь ласка, введіть додатнє ціле число.")
elif num == 0:
    print("Сума цифр числа 0 дорівнює 0.")
else:
    total_sum = 0
    temp_num = num
    while temp_num > 0:
        digit = temp_num % 10
        total_sum += digit
        temp_num = temp_num // 10
    print(f"Сума цифр числа {num} дорівнює {total_sum}.")

Введіть додатнє ціле число: 73
Сума цифр числа 73 дорівнює 10.


### Завдання 46. Найбільший спільний дільник ⭐⭐⭐

Запитайте у користувача два додатні числа і знайдіть їхній найбільший спільний дільник (НСД) за алгоритмом Евкліда.

**Алгоритм:** доки `b ≠ 0`: одночасно `a, b = b, a % b`. Коли `b = 0`, відповідь — це `a`.

**Приклад:** НСД(48, 36) = 12.

In [ ]:
a = int(input("Введіть перше додатнє число: "))
b = int(input("Введіть друге додатнє число: "))

if a <= 0 or b <= 0:
    print("Будь ласка, введіть додатні числа.")
else:
    while b != 0:
        a, b = b, a % b
    print(f"Найбільший спільний дільник: {a}")

Введіть перше додатнє число: 96
Введіть друге додатнє число: 33
Найбільший спільний дільник: 3


---

# Розділ 6. Списки (`list`) — методи та операції

**Що треба знати:**
- Створення: `[1, 2, 3]`, `[]` (порожній)
- Доступ: `lst[0]`, `lst[-1]` (останній)
- Довжина: `len(lst)`
- Методи: `.append(x)`, `.insert(i, x)`, `.remove(x)`, `.pop()`, `.sort()`, `.reverse()`
- Перевірка: `x in lst`

### Завдання 47. Створення списку ⭐

Створіть список з 5 ваших улюблених фільмів і виведіть його.

In [ ]:
favorite_movies = [
    "Фільм 1",
    "Фільм 2",
    "Фільм 3",
    "Фільм 4",
    "Фільм 5"
]
print("Мої улюблені фільми:")
print(favorite_movies)

Мої улюблені фільми:
['Фільм 1', 'Фільм 2', 'Фільм 3', 'Фільм 4', 'Фільм 5']


### Завдання 48. Додавання у список ⭐

Створіть порожній список. У циклі `for` додайте до нього числа від 1 до 10. Виведіть результат.

> 💡 **Підказка:** Використайте метод `.append()`.

In [ ]:
my_list = []
for i in range(1, 11):
    my_list.append(i)
print(my_list)

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]


### Завдання 49. Перший і останній ⭐

Дано список. Виведіть його перший і останній елементи.

In [ ]:
colors = ["червоний", "жовтий", "зелений", "синій", "фіолетовий"]
first_element = colors[0]
last_element = colors[-1]

print(f"Перший елемент: {first_element}")
print(f"Останній елемент: {last_element}")

Перший елемент: червоний
Останній елемент: фіолетовий


### Завдання 50. Список покупок ⭐⭐

Створіть список покупок з 3-4 товарів. Виконайте такі дії та виводьте список після кожної:
1. Додайте новий товар у кінець
2. Вставте новий товар на початок
3. Видаліть один товар за назвою
4. Замініть товар на іншому місці на щось нове

In [ ]:
shopping_list = ["молоко", "хліб", "яйця"]
print(f"Початковий список: {shopping_list}")

shopping_list.append("сир")
print(f"Після додавання сиру в кінець: {shopping_list}")

shopping_list.insert(0, "кава")
print(f"Після вставки кави на початок: {shopping_list}")

shopping_list.remove("хліб")
print(f"Після видалення хліба: {shopping_list}")

if "яйця" in shopping_list:
    index_of_eggs = shopping_list.index("яйця")
    shopping_list[index_of_eggs] = "масло"
print(f"Після заміни яєць на масло: {shopping_list}")

Початковий список: ['молоко', 'хліб', 'яйця']
Після додавання сиру в кінець: ['молоко', 'хліб', 'яйця', 'сир']
Після вставки кави на початок: ['кава', 'молоко', 'хліб', 'яйця', 'сир']
Після видалення хліба: ['кава', 'молоко', 'яйця', 'сир']
Після заміни яєць на масло: ['кава', 'молоко', 'масло', 'сир']


### Завдання 51. Сортування і реверс ⭐⭐

Дано список чисел. Відсортуйте його за зростанням, виведіть. Потім переверніть, виведіть.

In [ ]:
numbers = [42, 17, 89, 3, 56, 91, 24, 8, 73]

numbers.sort()
print(f"Відсортований список: {numbers}")

numbers.reverse()
print(f"Перевернутий список: {numbers}")

Відсортований список: [3, 8, 17, 24, 42, 56, 73, 89, 91]
Перевернутий список: [91, 89, 73, 56, 42, 24, 17, 8, 3]


### Завдання 52. Парні і непарні в окремі списки ⭐⭐

Дано список чисел. Створіть два нові списки: один з парними числами, другий з непарними. Виведіть обидва.

In [ ]:
numbers = [12, 7, 18, 3, 25, 9, 14, 21, 8, 11, 6, 33]

even_numbers = []
odd_numbers = []

for num in numbers:
    if num % 2 == 0:
        even_numbers.append(num)
    else:
        odd_numbers.append(num)

print(f"Парні числа: {even_numbers}")
print(f"Непарні числа: {odd_numbers}")

Парні числа: [12, 18, 14, 8, 6]
Непарні числа: [7, 3, 25, 9, 21, 11, 33]


### Завдання 53. Унікальні елементи ⭐⭐

Дано список з повторами. Створіть новий список, у якому кожен елемент зустрічається лише один раз.

> 💡 **Підказка:** Створіть порожній список `unique = []`. Перебирайте оригінальний список і додавайте елемент тільки якщо його ЩЕ НЕМАЄ в `unique`.

In [ ]:
items = ["яблуко", "банан", "яблуко", "груша", "банан", "слива", "яблуко", "груша"]
unique_items = []

for item in items:
    if item not in unique_items:
        unique_items.append(item)

print(f"Оригінальний список: {items}")
print(f"Унікальні елементи: {unique_items}")

Оригінальний список: ['яблуко', 'банан', 'яблуко', 'груша', 'банан', 'слива', 'яблуко', 'груша']
Унікальні елементи: ['яблуко', 'банан', 'груша', 'слива']


### Завдання 54. Підрахунок входжень ⭐⭐

Дано список і елемент. Підрахуйте, скільки разів цей елемент зустрічається у списку **без використання методу `.count()`**.

In [ ]:
words = ["кіт", "пес", "кіт", "хом'як", "кіт", "пес", "папуга", "кіт"]
target = "кіт"

count = 0
for item in words:
    if item == target:
        count += 1

print(f"Елемент '{target}' зустрічається {count} разів у списку.")

Елемент 'кіт' зустрічається 4 разів у списку.


### Завдання 55. Об'єднання списків ⭐⭐⭐

Дано два списки. Створіть новий список, який містить:
1. Спочатку елементи першого
2. Потім — елементи другого
3. Після — лише ті елементи з обох списків, які зустрічаються в обох

**Приклад:** `[1, 2, 3]` + `[2, 3, 4]` → `[1, 2, 3, 2, 3, 4, 2, 3]` (бо 2 і 3 — спільні).

In [ ]:
list_a = [1, 2, 3, 4, 5]
list_b = [3, 4, 5, 6, 7]

combined_list = []

for item in list_a:
    combined_list.append(item)

for item in list_b:
    combined_list.append(item)

for item_a in list_a:
    if item_a in list_b:
        combined_list.append(item_a)

print(f"Список A: {list_a}")
print(f"Список B: {list_b}")
print(f"Об'єднаний список з повторами спільних: {combined_list}")

Список A: [1, 2, 3, 4, 5]
Список B: [3, 4, 5, 6, 7]
Об'єднаний список з повторами спільних: [1, 2, 3, 4, 5, 3, 4, 5, 6, 7, 3, 4, 5]


### Завдання 56. Сортування без `.sort()` ⭐⭐⭐

Дано список чисел. Відсортуйте його за зростанням **без використання методу `.sort()` та функції `sorted()`**.

> 💡 **Підказка:** Найпростіший алгоритм — «бульбашкове сортування»: у вкладеному циклі порівнюємо сусідні елементи і міняємо їх місцями, якщо ліворуч стоїть більший. Повторюємо, поки не пройдемо весь список без жодної заміни.

In [ ]:
numbers = [42, 17, 89, 3, 56, 91, 24]
n = len(numbers)

for i in range(n):
    for j in range(0, n - i - 1):
        if numbers[j] > numbers[j + 1]:
            numbers[j], numbers[j + 1] = numbers[j + 1], numbers[j]

print(f"Відсортований список: {numbers}")

Відсортований список: [3, 17, 24, 42, 56, 89, 91]


---

# Розділ 7. 🔥 Комбіновані завдання

Тут вам знадобиться поєднати кілька різних тем у одному завданні. Це завдання, найбільш близькі до реальних задач, які доводиться розв'язувати програмістам.

### Завдання 57. Меню кафе ⭐⭐

Дано список страв і список їхніх цін (відповідні позиції).
1. Виведіть меню у форматі «Назва — Ціна грн».
2. Підрахуйте і виведіть середню ціну.
3. Виведіть страви, ціна яких вища за середню.

> 💡 **Підказка:** Щоб одночасно перебирати два списки, використовуйте індекс: `for i in range(len(dishes)):`, далі `dishes[i]` і `prices[i]`.

In [ ]:
dishes = ["Борщ", "Вареники", "Салат", "Стейк", "Десерт", "Чай"]
prices = [85, 95, 70, 320, 110, 35]

print("Меню:")
for i in range(len(dishes)):
    print(f"{dishes[i]} — {prices[i]} грн")

total_price = sum(prices)
average_price = total_price / len(prices)
print(f"\nСередня ціна страви: {average_price:.2f} грн")

print("\nСтрави, ціна яких вища за середню:")
for i in range(len(dishes)):
    if prices[i] > average_price:
        print(f"{dishes[i]} — {prices[i]} грн")

Меню:
Борщ — 85 грн
Вареники — 95 грн
Салат — 70 грн
Стейк — 320 грн
Десерт — 110 грн
Чай — 35 грн

Середня ціна страви: 119.17 грн

Страви, ціна яких вища за середню:
Стейк — 320 грн


### Завдання 58. Шифр Цезаря ⭐⭐⭐

Запитайте у користувача слово (українською або англійською) і число `shift`.
Зашифруйте слово: кожну літеру замініть на ту, що стоїть на `shift` позицій далі в алфавіті.

**Приклад:** слово `abc`, `shift = 2` → `cde`.

> 💡 **Підказка:** Використайте функції `ord(letter)` (повертає код символу) і `chr(num)` (повертає символ за кодом). Для кожної літери: новий код = ord(letter) + shift. Для простоти можна не обробляти переходи через кінець алфавіту.

In [ ]:
word = input("Введіть слово для шифрування: ")
shift = int(input("Введіть зсув (число): "))

encrypted_word = ""
for char in word:
    if 'a' <= char <= 'z':
        start = ord('a')
        encrypted_char_code = (ord(char) - start + shift) % 26 + start
        encrypted_word += chr(encrypted_char_code)
    elif 'A' <= char <= 'Z':
        start = ord('A')
        encrypted_char_code = (ord(char) - start + shift) % 26 + start
        encrypted_word += chr(encrypted_char_code)
    elif 'а' <= char <= 'я':
        encrypted_word += chr(ord(char) + shift)
    elif 'А' <= char <= 'Я':
        encrypted_word += chr(ord(char) + shift)
    else:
        encrypted_word += char

print(f"Зашифроване слово: {encrypted_word}")

Введіть слово для шифрування: world
Введіть зсув (число): 3
Зашифроване слово: zruog


### Завдання 59. Гра «Вгадай число» ⭐⭐⭐

Створіть змінну `secret = 42` (загадане число). У циклі `while` запитуйте у користувача число:
- Якщо введене число менше за загадане — пишіть «Більше»
- Якщо більше — «Менше»
- Якщо вгадав — пишіть «Вітаю!» і завершуйте цикл
- Підрахуйте кількість спроб і виведіть її в кінці

> 💡 **Підказка:** Використайте змінну-прапорець `guessed = False` як умову циклу. Коли користувач вгадав — `guessed = True`.

In [ ]:
secret = 42
attempts = 0
guessed = False

print("Я загадав число від 1 до 100. Спробуй вгадати!")

while not guessed:
    try:
        guess = int(input("Введіть ваше число: "))
        attempts += 1

        if guess < secret:
            print("Більше")
        elif guess > secret:
            print("Менше")
        else:
            print(f"Вітаю! Ви вгадали число {secret} за {attempts} спроб!")
            guessed = True
    except ValueError:
        print("Будь ласка, введіть ціле число.")

Я загадав число від 1 до 100. Спробуй вгадати!
Введіть ваше число: 12
Більше
Введіть ваше число: 92
Менше
Введіть ваше число: 42
Вітаю! Ви вгадали число 42 за 3 спроб!


### Завдання 60. Аналіз тексту ⭐⭐⭐

Дано рядок з реченням. Виконайте такий аналіз і виведіть результати:
1. Кількість літер (без пробілів)
2. Кількість слів
3. Кількість голосних літер
4. Найдовше слово

> 💡 **Підказка:** Метод `.split()` розбиває рядок на список слів. Наприклад, `'hello world'.split()` повертає `['hello', 'world']`.

In [ ]:
text = "програмування на python це цікаво і корисно"

letter_count = 0
for char in text:
    if char.isalpha():
        letter_count += 1

words = text.split()
word_count = len(words)

vowels = "аеиіоуяюєїaeiou"
vowel_in_text_count = 0
for char in text.lower():
    if char in vowels:
        vowel_in_text_count += 1

longest_word = ""
if words:
    longest_word = words[0]
    for word in words:
        if len(word) > len(longest_word):
            longest_word = word

print(f"Текст для аналізу: '{text}'")
print(f"1. Кількість літер (без пробілів): {letter_count}")
print(f"2. Кількість слів: {word_count}")
print(f"3. Кількість голосних літер: {vowel_in_text_count}")
print(f"4. Найдовше слово: '{longest_word}'")

Текст для аналізу: 'програмування на python це цікаво і корисно'
1. Кількість літер (без пробілів): 37
2. Кількість слів: 7
3. Кількість голосних літер: 15
4. Найдовше слово: 'програмування'


### Завдання 61. Магазин: статистика продажів ⭐⭐⭐

Дано список продажів за тиждень (по одному числу на день, починаючи з понеділка).
Виведіть:
1. Загальну виручку за тиждень
2. Середню виручку на день
3. Найкращий день (з найбільшою виручкою) — назвою (Понеділок, Вівторок...)
4. Кількість днів, коли виручка була вищою за середню

> 💡 **Підказка:** Створіть список з назвами днів: `days = ["Понеділок", "Вівторок", ...]`. Індекс найкращого дня знайдіть у циклі (або через `sales.index(max(sales))`).

In [ ]:
sales = [3200, 2800, 4100, 3500, 5200, 7800, 6500]
days_of_week = ["Понеділок", "Вівторок", "Середа", "Четвер", "П'ятниця", "Субота", "Неділя"]

total_revenue = sum(sales)
print(f"1. Загальна виручка за тиждень: {total_revenue} грн")

average_revenue = total_revenue / len(sales)
print(f"2. Середня виручка на день: {average_revenue:.2f} грн")

max_sales = 0
best_day_index = -1
for i in range(len(sales)):
    if sales[i] > max_sales:
        max_sales = sales[i]
        best_day_index = i

if best_day_index != -1:
    print(f"3. Найкращий день: {days_of_week[best_day_index]} ({max_sales} грн)")
else:
    print("3. Не вдалося визначити найкращий день.")

days_above_average = 0
for s in sales:
    if s > average_revenue:
        days_above_average += 1
print(f"4. Кількість днів з виручкою вищою за середню: {days_above_average}")

1. Загальна виручка за тиждень: 33100 грн
2. Середня виручка на день: 4728.57 грн
3. Найкращий день: Субота (7800 грн)
4. Кількість днів з виручкою вищою за середню: 3


### Завдання 62. Турнірна таблиця ⭐⭐⭐

Дано два списки: команди та їхні очки (відповідні позиції).
Виведіть турнірну таблицю, ВІДСОРТОВАНУ за спаданням очок, у такому форматі:
```
1. Шахтар — 67 очок
2. Динамо — 64 очки
...
```

> 💡 **Підказка:** Найпростіший спосіб: у циклі знаходьте команду з максимальними очками, виводьте її, потім видаляйте цю пару (команду і очки) з обох списків. Повторюйте, поки списки не спорожніють.

In [ ]:
teams = ["Динамо", "Шахтар", "Зоря", "Дніпро", "Ворскла", "Олександрія"]
points = [64, 67, 52, 48, 39, 45]
print("Турнірна таблиця:")

current_teams = list(teams)
current_points = list(points)

sorted_table_data = []

while current_teams:
    max_points_current_iter = -1
    max_index_current_iter = -1

    for i in range(len(current_points)):
        if current_points[i] > max_points_current_iter:
            max_points_current_iter = current_points[i]
            max_index_current_iter = i

    if max_index_current_iter != -1:
        team_name = current_teams.pop(max_index_current_iter)
        team_points = current_points.pop(max_index_current_iter)
        sorted_table_data.append((team_name, team_points))

for rank, (team, score) in enumerate(sorted_table_data, 1):
    print(f"{rank}. {team} — {score} очок")

Турнірна таблиця:
1. Шахтар — 67 очок
2. Динамо — 64 очок
3. Зоря — 52 очок
4. Дніпро — 48 очок
5. Олександрія — 45 очок
6. Ворскла — 39 очок


---

# 🎉 Молодці!

Якщо ви виконали всі завдання — ви великі молодці! Це справді багато практики, і кожне завдання робило вас сильнішим програмістом.

### 💡 Що робити далі

- Якщо щось залишилось незрозумілим — поверніться до того завдання і спробуйте ще раз
- Перепишіть розв'язки складних завдань з пам'яті — це найкращий спосіб закріпити
- Спробуйте придумати свої варіації цих задач і розв'язати їх

**Не зупиняйтесь — програмування вчиться лише через практику! 🐍**
